File: App/Resources/AlgorithmDetails/bitonicsortrecursive/bitonicsortrecursive.rb
def greatest_power_of_two_less_than(n)
  k = 1
  while k < n
    k <<= 1
  end
  k >> 1
end

def compare(array, i, j, dir)
  is_greater = array[i] > array[j]
  if dir == is_greater
    array[i], array[j] = array[j], array[i]
  end
end

def bitonic_merge(array, lo, n, dir)
  if n > 1
    m = greatest_power_of_two_less_than(n)
    (lo...(lo + n - m)).each do |i|
      compare(array, i, i + m, dir)
    end
    bitonic_merge(array, lo, m, dir)
    bitonic_merge(array, lo + m, n - m, dir)
  end
end

def bitonic_sort(array, lo, n, dir)
  if n > 1
    m = n / 2
    bitonic_sort(array, lo, m, !dir)
    bitonic_sort(array, lo + m, n - m, dir)
    bitonic_merge(array, lo, n, dir)
  end
end

def sort(array)
  bitonic_sort(array, 0, array.length, true)
end

array = [0, 39, 21, 62, 91, 77, 14, 23,
  90, 69, 51, 81, 68, 83, 32, 56]
sort(array)
p array

File: App/Resources/AlgorithmDetails/bitonicsortrecursive/bitonicsortrecursive.java
import java.util.Arrays;

public class bitonicsortrecursive {
  private static int greatestPowerOfTwoLessThan(int n) {
    int k = 1;
    while (k < n) {
      k <<= 1;
    }
    return k >> 1;
  }

  private static void compare(int[] array, int i, int j, boolean dir) {
    boolean isGreater = array[i] > array[j];
    if (dir == isGreater) {
      int t = array[i];
      array[i] = array[j];
      array[j] = t;
    }
  }

  private static void bitonicMerge(int[] array, int lo, int n, boolean dir) {
    if (n > 1) {
      int m = greatestPowerOfTwoLessThan(n);
      for (int i = lo; i < lo + n - m; i++) {
        compare(array, i, i + m, dir);
      }
      bitonicMerge(array, lo, m, dir);
      bitonicMerge(array, lo + m, n - m, dir);
    }
  }

  private static void bitonicSort(int[] array, int lo, int n, boolean dir) {
    if (n > 1) {
      int m = n / 2;
      bitonicSort(array, lo, m, !dir);
      bitonicSort(array, lo + m, n - m, dir);
      bitonicMerge(array, lo, n, dir);
    }
  }

  public static void sort(int[] array) {
    bitonicSort(array, 0, array.length, true);
  }

  public static void main(String[] args) {
    int[] array = new int[] {0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56};
    sort(array);
    System.out.println(Arrays.toString(array));
  }
}
